Make code actions available in DafnyCore

keyboardDrummer opened the pull request:
### Changes
- Merge the classes `ErrorDetail` and `DafnyCodeActions` into `ErrorRegistry`, which is located in DafnyCore. As such, error descriptions and code actions can be registered from within DafnyCore. This allows:
  - Creating a command `dafny format --code-actions` which applies code actions in addition to formatting`
  - Registering code actions for errors in the same locations as where the errors are created.
- `DiagnosticDafnyCodeActionProvider.GetDafnyCodeActions` now takes a `DafnyDiagnostic` that has an `IToken`, which makes it so that no ranges have to be passed back and forth in the `Data` field of code actions.
- Added an alternative to `TestCodeActionHelper`, `TestCodeAction` that allows specifying the range in which the action replaces things, which was needed for one of the new tests. @MikaelMayer, it might be good to unify `TestCodeAction` and `TestCodeActionHelper`. Could you comment on the markup syntax that I used for `TestCodeAction`? It's a little ugly but it does reuse `MarkupTestFile`.
- Add `TODO` markers for various errors which are registered in `ParseErrors` but which are never reported by the Parser. This is effectively dead code. 

### Testing
- Add a test for the code action that removes an illegal underscore
- Add a test for the code action that removes a deprecated attribute

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

MikaelMayer commented on Source/DafnyCore/Dafny.atg:
```diff
@@ -2008,7 +2007,7 @@ MethodDecl<DeclModifierData dmod, bool allowConstructor, out Method/*!*/ m>
   (. if (!hasName) {
        if (!isConstructor) {
-         SemErr(ErrorID.p_method_missing_name, la, "a method must be given a name (expecting identifier)");
+         SemErr(ParseErrors.ErrorId.p_method_missing_name, la, "a method must be given a name (expecting identifier)");
```
Could you import ErrorId so that you don't need to repeat ParseErrors.ErrorId, just ErrorId? You could also consider renaming ParseErrorId if you report things other than parse errors.

MikaelMayer reviewed (changes requested):
I love the idea of enhancing the formatter to support code actions !
A few suggestions to enhance this PR while you are working on it.

davidcok commented:
OK. Here is what I take as the changes so far in this refactoring.

* Name changes: ID to Id (both are used for identifier), ErrorDetail to ErrorRegistry, ParserErrorDetail to ParseErrors, etc. These mostly seem gratuitous, but I'm not going to quibble. Just that doing name changes as a separate PR would have significantly reduced the bulk of a PR that had more substantive changes.
* Moving ErrorIds from one global file into a file associated with the area of code in which those errors occur. I like the locality. The downside is that the ErrorIds are not all the same type and have to devolve to strings earlier than they would otherwise. However, that is hidden from the point of the error message and does not have any performance problems, so this refactoring is fine with me. Also unfortunately, it makes things more verbose as ErrorId is now ParseErrors.ErrorId throughout Dafny.atg.
* The other aspect of that is that ErrorId.None is replaced by null. ErrorId.None served three purposes: it was clear on the caller side that this parameter is supposed to be an ErrorId, it avoided some ambiguities that simply using null caused (when there were other default parameters), and it was a signal (that could easily be found by searching) to the developer (me initially, others later) that there was no ErrorId defined for this error message. I still think those are valuable and would like to retain ErrorId.None, even if it means having a local method that takes an ErrorId and then immediately calls Reporter where a string is needed.
* You changed the lazy init method to a static constructor. I initially tried a static constructor, but it seemed to not work reliably in testing. Not having experience with that aspect of C#, and the fact that the test failures went away when I changed away from static constructors, led to the lazy initialization. But I'm completely open to a robust static-constructor solution. Plus those problems I experienced might have been conflated with the other testing problems we were having at the same time.
* The change of location of Printer.cs and TokenNewIdentCollector.cs - does that change really belong in this refactoring?
* All the changes to Tokens and Ranges - I was expecting this. I'll have to look at it in more detail to understand its implications here, but I'm willing to accept that you've worked out the various use cases for this -- and am glad you've worked through the changes for these code actions. There are quite a few changes (improvements) to error messages in the 4.0 code that are, unfortunately, going to conflict (on merging) with these changes. Not insurmountably.
* One aspect of the ranges changes -- using just a parser token rather than what was a CodeActionRange -- makes it less clear at the call site which range is used for display and which is used for code action. I'll have to think about that aspect more, once I see this merged in and work with it.
* Putting the code action definition with the error id is a good thing (plus the potential ability to apply the action programmatically outside the LSP) -- I did not like the separation in the original design. But the reason for the separation was to avoid pulling LSP code into the core of dafny. I see that you've encapsulated the action in a DafnyCodeActionEdit record, which suffices for the simple range replacement edits that were defined so far. But there will be more complex changes as well, so I suspect this structure will need expansion.

keyboardDrummer commented on Source/DafnyLanguageServer.Test/Synchronization/OpenDocumentTest.cs:
```diff
@@ -80,9 +80,9 @@ public class OpenDocumentTest : DafnyLanguageServerTestBase {
-      var message = document.Diagnostics.First(d => d.Severity!.Value == DiagnosticSeverity.Error);
-      Assert.AreEqual(MessageSource.Verifier.ToString(), message.Source);
+      Assert.AreEqual(1, document.Diagnostics.Count(d => d.Level == ErrorLevel.Error));
+      var message = document.Diagnostics.First(d => d.Level == ErrorLevel.Error);
```
Good suggestion, I'll do it in a follow-up PR.

keyboardDrummer commented on Source/DafnyCore/AST/Tokens.cs:
```diff
@@ -244,16 +250,13 @@ public class BoogieRangeToken : TokenWrapper {
-    this.column = column;
-    this.length = length;
-  }
+public class TokenRange : TokenWrapper {
```
> Can you please justify why we have "RangeToken" and "TokenRange" which both contain a start and end token? I know that "TokenRange" is a TokenWrapper whereas "RangeToken" is not, but I feel that they are too close to each other so that having both has a benefit.

Removed one.

> Moreover, previously, it was possible to distinguish code actions that included or excluded the last token, which seems not encoded anymore.

I've changed the semantics of RangeToken so that if EndToken is null, it's only the starting token that defines the range, which seems sufficient.

MikaelMayer replied:
Can you please reassure me?
Until before this PR, a RangeToken was always inclusive from the start of the StartToken to the end of the EndToken. This is what makes it possible to highlight in the IDE an entire expression.

From what you say, it looks like the new semantics are that, if the RangeToken does not have an end, then the startToken defines the beginning and the end inclusive, but if it has an end, then the end of the range is the start of the EndToken
Am I missing something? If this is true, how do you make distinction between the two?

keyboardDrummer replied:
If there's an StartToken and EndToken, then the semantics are unchanged: the range will include the whole EndToken.

keyboardDrummer commented on Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs:
```diff
@@ -64,7 +64,7 @@ public class ClientBasedLanguageServerTest : DafnyLanguageServerTestBase {
     do {
       result = await diagnosticsReceiver.AwaitNextDiagnosticsAsync(cancellationToken);
-    } while (!document!.Diagnostics.SequenceEqual(result));
+    } while (!document!.Diagnostics.Select(d => d.ToLspDiagnostic()).SequenceEqual(result));
```
To compare apples to apples. This method is comparing internal state (DafnyDiagnostic) with external state (LSP Diagnostic), and for that it first needs to convert the internal state to the same type as the external one.

keyboardDrummer commented on Source/DafnyCore/Generic/BatchErrorReporter.cs:
```diff
@@ -0,0 +1,35 @@
+using System.Collections.Generic;
```
Note there's things in this folder unrelated to error reporting. The goal was to collect things that don't relate to Dafny in a particular folder. Within that again there's a most generic category like `SingleLinkedList` or `StrongConnectedComponent`, and a `Compilers` category.

keyboardDrummer commented:
@davidcok it might be easier to reply to your comment if you create separate comments for the various points, so there are multiple threads. 
> 
> * Name changes: ID to Id (both are used for identifier), ErrorDetail to ErrorRegistry, ParserErrorDetail to ParseErrors, etc. These mostly seem gratuitous, but I'm not going to quibble. Just that doing name changes as a separate PR would have significantly reduced the bulk of a PR that had more substantive changes.

ErrorDetail has been merged with a part of DafnyCodeActions so it's no longer only about details/descriptions of errors, so it seemed good to update the name. Similarly, ParserErrorDetails has also merged with a part of DafnyCodeActions.

> * Moving ErrorIds from one global file into a file associated with the area of code in which those errors occur. I like the locality. The downside is that the ErrorIds are not all the same type and have to devolve to strings earlier than they would otherwise. However, that is hidden from the point of the error message and does not have any performance problems, so this refactoring is fine with me. 

Great :-)

> Also unfortunately, it makes things more verbose as ErrorId is now ParseErrors.ErrorId throughout Dafny.atg.

Resolved.

> * The other aspect of that is that ErrorId.None is replaced by null. ErrorId.None served three purposes: it was clear on the caller side that this parameter is supposed to be an ErrorId, it avoided some ambiguities that simply using null caused (when there were other default parameters), and it was a signal (that could easily be found by searching) to the developer (me initially, others later) that there was no ErrorId defined for this error message. I still think those are valuable and would like to retain ErrorId.None, even if it means having a local method that takes an ErrorId and then immediately calls Reporter where a string is needed.

Resolved

> * You changed the lazy init method to a static constructor. I initially tried a static constructor, but it seemed to not work reliably in testing. Not having experience with that aspect of C#, and the fact that the test failures went away when I changed away from static constructors, led to the lazy initialization. But I'm completely open to a robust static-constructor solution. Plus those problems I experienced might have been conflated with the other testing problems we were having at the same time.

Static constructors are lazy, only run once, and are thread-safe.

> * The change of location of Printer.cs and TokenNewIdentCollector.cs - does that change really belong in this refactoring?

I had to give a location to ParseErrors, for which I created a new folder `AST/Grammar`, and I felt it would be good to flesh out this folder somewhat with the files that belong it in. Moving unchanged files doesn't inflate the PR diff much so that seems like a cheap thing to do. I've moved a few more files into it since your comment.

> * One aspect of the ranges changes -- using just a parser token rather than what was a CodeActionRange -- makes it less clear at the call site which range is used for display and which is used for code action. I'll have to think about that aspect more, once I see this merged in and work with it.

The entire range is potentially used for display, although currently I think we're only displaying the start of it. It's up to the code action handler to infer a code action range from the display range. For our actions, those two often end up being the same.

I think a more powerful API would be that at the site where you report an error, you also create the entire code action, which is then stored together with the error and can be used to return a code action when the user invokes code action on that diagnostic. However, it seems we don't need that yet.

> * Putting the code action definition with the error id is a good thing (plus the potential ability to apply the action programmatically outside the LSP) -- I did not like the separation in the original design. But the reason for the separation was to avoid pulling LSP code into the core of dafny. I see that you've encapsulated the action in a DafnyCodeActionEdit record, which suffices for the simple range replacement edits that were defined so far. But there will be more complex changes as well, so I suspect this structure will need expansion.

I expect that as well but I think it's better to do such an expansion only when we need it.

MikaelMayer commented on Source/DafnyLanguageServer.Test/CodeActions/CodeActionsTest.cs:
```diff
@@ -26,6 +26,23 @@ public class CodeActionTest : ClientBasedLanguageServerTest {
+      await TestCodeAction(@"
+method Foo()
+{
+  var ><|>remove underscore>>>x|||_x<| := 3; 
```
I like your new language to encode code actions on top of a Dafny file, very clever. As you probably saw in the rest of the file, there was another language to do that, however, it did not have support yet for erasing things, only adding things.
Did you look into expanding that previous language? Your example could have look like this:
```suggestion
  var >>>[[remove underscore|x<<<_x]] := 3; 
```

Otherwise, if you want to keep the spirit of your language, I have a suggestion to increase its legibility
```suggestion
  var ><(>remove underscore|x<<<_x<) := 3; 
```
* Parentheses so that we visually group the entire code action.
* Similar to previous `[>` and `{>`
* The new separator after the label (`|`) does not need to be complicated since label probably won't contain that separator symbol (and if it does later, we could always change it later)
* Order of `<<<` indicates what is old (on the right of the arrow) vs. what is new (on the left of the arrow)
* Visually, `>` binds stronger than `|`, so now there is the label on one side and the transformation on the other side.

keyboardDrummer replied:
> Parentheses so that we visually group the entire code action.

Good idea. Done

> Did you look into expanding that previous language?

I wanted to re-use the existing `MarkupTestFile` language, since that's also used outside of code actions. I think the `><` that we have there is good for indicating a position, so while the `>>>` makes more sense for code actions since the position is followed by an action, I think I'll keep the `><` for consistency.

> Order of <<< indicates what is old (on the right of the arrow) vs. what is new (on the left of the arrow)

Makes sense, but this operator is also the generic binding operator for metadata and code in `MarkupTestFile`. I've change dit from `|||` to `:::` since that seems more like a 'binding' operator.

---

If you want, I can also update the other tests to use this, arguably less nice language than the existing one.

MikaelMayer replied:
Nice, I find that it now reads nicely, especially when you can omit the position in front of the range.
Yes, I'd love that you update the other tests to use this _now arguably nice_ markup language, so that we use only one language. Easier to maintain.

keyboardDrummer replied:
Done

MikaelMayer reviewed (commented):
This is in good shape ! I think I have two suggestions/questions to sort out and we should be good to go !

MikaelMayer commented on Source/DafnyCore/AST/Grammar/ParseErrors.cs:
```diff
@@ -126,7 +126,7 @@ public enum ErrorId {
 ", range => new List<DafnyAction> { new("remove underscore",
     new [] {
-      new DafnyCodeActionEdit(range, range.PrintOriginal().Substring(1))
+      new DafnyCodeActionEdit(range.ToDafnyRange(), range.PrintOriginal().Substring(1))
```
Could you perhaps add an overload of the constructor so that you don't need to call ToDafnyRange on the first argument? Seems like you do this a lot.

MikaelMayer reviewed (approved):
Looks good to me ! Thanks for adding all the tests
